src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketPullRequestFull.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016-2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.cloudbees.jenkins.plugins.bitbucket.api;
+
+import java.util.List;
+
+/**
+ * Represents all the details of a pull request.
+ */
+public interface BitbucketPullRequestFull extends BitbucketPullRequest {
```

Comment:
Is this really necessary? Why not just add the `getReviewers` to the `BitbucketPullRequest` class

Reply:
Yeah, I think it could be added to the existing mapping class.

Reply:
I totally see how this could look weird so let me explain. 

Right now the plugin pulls in all the pull requests for a particular repo with a url such as:
`https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests{?page,pagelen}`
This returns all the data contained in the `BitbucketPullRequest` interface for every PR in that repo. However, it does not return any data about reviewers/approvers in that response. In order to get that you need to call a different endpoint along the lines of: 
`https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests/{id}` (notice that we are accessing a single PR ID)
This is the endpoint that the `getPullRequestById()` function calls and thus includes the reviewer information. (This function is never called within this plugin as far as I'm aware currently.)
Taking this all into account I created the new interface so that everything that currently exists would have the same interface to interact with and not have to worry that `getReviewers()` would always return null and `getPullRequestById()` could return the full version since it does have access to all the data. 

Hopefully the reasoning is a bit clearer now. I definitely wish I didn't have to do this but alternatives were along the lines of making an additional HTTP call for each pull request to get the additional data and that seemed much worse.

Reply:
Ok, so we need a new HTTP call. Still I would keep the client API object _hidden_, my proposal:

1. Add `BitbucketSCMSourceRequest#getPullRequestById(String id)`
1. It will use [`source.buildBitbucketClient()`](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/fc0785417c4cdc6afc777f88c2447cfeb8266f2c/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java#L508) to create a client and make the HTTP call to get the PR details - note that the `BitbucketSCMSource` is [passed to the request](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/2379c9b5a581d09688567e45de66b4f86f822de8/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java#L126).
1. Use `BitbucketSCMSourceRequest#getPullRequestById(String id)` in the extension plugin.
1. The implementation of `BitbucketSCMSourceRequest#getPullRequestById(String id)` should use some sort of caching mechanism. Two consecutive calls to `getPullRequestById` for the same PR should do no more than one HTTP request. The lifecycle of the `BitbucketSCMSourceRequest` object is bounded to a specific `SCMSource#retrieve` call so the cache won't need any _invalidate_, _erase_ nor _eviction_ handling (a plain old `HashMap<String, BitbucketPullRequestFull>` should be enough.

This way `BitbucketApi` remains _private_ and extensions can get PR details by using the already available `BitbucketSCMSourceRequest` object.

WDTY?

Reply:
Seems reasonable to me. I'll implement that when I get a chance and update here. Thank you for the suggestion.

Reply:
Until we get java 8 adding more interfaces just for a new method is dangerous, what to do next time we want to add a new method? New interface? what should it be called? `BitbucketPullRequestFuller`? :)
Since we don't have default methods yet, maybe a new interface is needed, but it could simply be called `BitbucketPullRequest2`. wdyt?

Reply:
So the reasoning behind the name is, as somewhat explained above, there are two sources of pull request metadata. The first source, and the one primarily used in this plugin, is the `https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests{?page,pagelen}` endpoint. This of course gets data about multiple pull requests at a time which leads to a reduced number of HTTP requests required to get the data. However this endpoint only returns a subset of the data about a pull request, one of the pieces it doesn't include is reviewer information. That comes only from the other source of pull request metadata which is `https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests/{id}` this only returns data about one pull request but includes all the data about the pull request. This is why I chose the name, because this is the place where you get the full information about a pull request. So if we had a new method we wanted to add and the source of it's data was not one of the above two endpoints we could indeed be in an annoying situation with naming. My bet with the naming of this was that it wouldn't happen though. 
I'm fine to change the name if there is passion behind it but I personally like `BitbucketPullRequestFull` the most of any option I have thought of.

Reply:
Still don't think is needed: You could settle for `@JsonInclude(Include.NON_EMPTY)` on `getReviewers`